### PR TITLE
Update AOTAutograd to use FunctionalTensorMode instead of C++ functionalization

### PR DESCRIPTION
**Pull request by bdhirsh**
Now that FunctionalTensor and `FunctionalTensorMode` are lower down in this stack, the changes in this PR are more mechanical: Everywhere in AOTAutograd that I used to use the C++ functionalization API, I now use the python functionalization API.

Note that this doesn't actually cause functionalization to run underneath torch_dispatch. I'm saving that re-ordering for later in the stack.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105236
* #104974
* #104483
* __->__ #106406
* #109023
* #109632
* #109662
* #108654

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by bdhirsh** [comment]
Most of the tests that are failing are on higher order ops - it looks like they've all been given functionalization rules that are reigstered directly to the `Functionalize` key, but they'll need to be updated to work with `FunctionalTensorMode`.

This might require updating them not to use `functorch.functionalize`, hmm (cc @zou3519).

**Comment by bdhirsh** [comment]
Higher order ops <> FunctionalTensorMode interaction should now be fixed.

Hmmm, ASAN is failing (causing the aot autograd tests to fail when I turn FunctionalTensorMode on)  - looks like it has to do with the `PyInterpreter::sym_sizes()` code for stashing symints:
```
  
  0x60300117a790 is located 16 bytes inside of 32-byte region [0x60300117a780,0x60300117a7a0)
  freed by thread T0 here:
      #0 0x7f411a89323d in operator delete[](void*) (/usr/lib/llvm-12/lib/clang/12.0.1/lib/linux/libclang_rt.asan-x86_64.so+0xce23d)
      #1 0x7f4112105648 in (anonymous namespace)::ConcretePyInterpreterVTable::sym_sizes(c10::TensorImpl const*) const::$_3::__invoke(void*) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_python.so+0x258b648)
      #2 0x7f41117fb4aa in pybind11::capsule::initialize_with_void_ptr_destructor(void const*, char const*, void (*)(void*))::'lambda'(_object*)::operator()(_object*) const (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_python.so+0x1c814aa)
      #3 0x4573da in capsule_dealloc.cold /usr/local/src/conda/python-3.9.17/Objects/capsule.c:261:9
  
  previously allocated by thread T0 here:
      #0 0x7f411a8929ed in operator new[](unsigned long) (/usr/lib/llvm-12/lib/clang/12.0.1/lib/linux/libclang_rt.asan-x86_64.so+0xcd9ed)
      #1 0x7f41120fa60d in (anonymous namespace)::ConcretePyInterpreterVTable::sym_sizes(c10::TensorImpl const*) const (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_python.so+0x258060d)
      #2 0x7f40ffc010cb in torch::autograd::generated::details::fft_r2c_backward(at::Tensor const&, c10::ArrayRef<long>, long, bool, c10::SymInt) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0x131910cb)
      #3 0x7f40fbfc4548 in torch::autograd::generated::FftR2CBackward0::apply(std::vector<at::Tensor, std::allocator<at::Tensor> >&&) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0xf554548)
      #4 0x7f40fe414a80 in torch::autograd::Node::operator()(std::vector<at::Tensor, std::allocator<at::Tensor> >&&) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0x119a4a80)
      #5 0x7f40fe402eb3 in torch::autograd::call_function(std::shared_ptr<torch::autograd::GraphTask>&, torch::autograd::Node*, torch::autograd::InputBuffer&) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0x11992eb3)
      #6 0x7f40fe3fac99 in torch::autograd::Engine::evaluate_function(std::shared_ptr<torch::autograd::GraphTask>&, torch::autograd::Node*, torch::autograd::InputBuffer&, std::shared_ptr<torch::autograd::ReadyQueue> const&) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0x1198ac99)
      #7 0x7f40fe3f918c in torch::autograd::Engine::thread_main(std::shared_ptr<torch::autograd::GraphTask> const&) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0x1198918c)
```

**Comment by soulitzer** [comment]
I think its happening because we're doing something like:
```
c10::SymIntArrayRef x = a.sym_sizes();  // writes to the _sym_sizes_capsule field
c10::SymIntArrayRef  = a.sym_sizes();  // overwrites the existing field, and x points to freed memory!
```
We should check if the tensor already has a _sym_sizes_capsule field, and return SymIntArrayRef to the existing one if it exists. 
But maybe we don't actually want to cache this value in case it changes.

In that case, we would want to increment an counter, so that every time we access sym_size we write to a different field. Hopefully people don't call sym_sizes too many times on the same exact tensor lol

**Comment by bdhirsh** [comment]
Good catch - thanks @soulitzer!

Hmm @ezyang what do you think? I don't really see any options other than:

(1) allocate a fresh buffer that we stash each time `sym_size()` is called (feels bad - if we did this, we should probably run some example models/tests and confirm that we aren't actually allocating hundreds of buffers per tensor)

(2) Go back to the caching strategy from before. This is a non-starter, since it did not handle metadata mutations

(3) Cache but handle metadata mutations (so we only allocate a fresh buffer when we figure out that `sym_size()` was called after a metadata mutation).

My feeling is that it might end up being necessary to do (3) for `FunctionalTensor`, since it'll be used everywhere - but it might be a bit much to require all subclasses to do this by default (so (2) can be the default).

**Comment by bdhirsh** [comment]
Concretely, I think we can do (3) in `FunctionalTensor` by storing a counter on the subclass (like Jeffrey mentioned), that we increment every time we see a metadata mutation op. Calls to `sym_size()` will try to grab a cached value for `__sym_size_buffer{counter}` if it exists, and otherwise set it.

**Comment by bdhirsh** [comment]
Segfault is fixed locally (although a few expect tests in test_aotdispatch.py are failing, still looking into those).

So far locally, I did an even dumber version of (3): if the subclass doesn't make sure to set a `metadata_mutation_ctr` attribute on their subclass properly, then I will just always re-use the first cached buffer (so you will always get a stale value if your tensor has mutated metadata).

This still seems like a strict improvement over the current state - in my head, this feeds into the larger question from https://github.com/pytorch/pytorch/pull/107915 around "what can we do to ensure that a subclass author has to do as little as possible to get metadata / mutation correctness? I'll push my current changes once I get more tests passing, but if we think that we should fix this issue in a better way in this PR, happy to do it.

**Comment by bdhirsh** [comment]
For the race conditions: It might also be simpler if we can just require the subclass to invoke some code that fixes up the buffer whenever a metadata mutation happens) so we can guarantee that when a user calls `.size/stride`, the buffer is set up correctly and no mutation is necessary).

This would mean more requirements on the subclass for correctness though.

**Comment by ezyang** [comment]
You don't need to store all of the old ones. If you triggered a resize from 5 to 6, old IntArrayRef is going to be invalidated too. We are primarily trying to make sure the no-op case works fine. The higher fidelity is mostly for bug-for-bug compatibility in case someone is doing something naughty, I don't actually expect it to matter in practice.

Forget about the high fidelity thing for a moment. Concretely, this is what I propose you do:

1. You will maintain a single buffer, sized to exactly what the reported dim of the tensor is. The invariant will be that we will preserve ArrayRef only if there is no change in the reported size. If there is a resize, all old ArrayRef are invalidated, multithreaded access at the same time as resize is also a no-no.
2. When a user accesses `sym_size` in C++, we will make the call to Python. If the dim of the reported size is different from the buffer, we resize the buffer (deleting the old buffer and invalidating old ArrayRef). If resized, we write in all elements. Otherwise, for each element, we test if it has changed. If it has changed, we write the updated element in. (It probably is safe to unconditionally write the elements in but doing the tests makes us technically data race free; without the tests we are technically not data race free because we have an unprotected read-write race (even though the write is a no-op). Some care needs to be taken about ownership for SymInt.

**Comment by ezyang** [comment]
In TensorImpl, assuming you are > 5, we always reallocate the memory even if you get smaller. This allows to avoid recording an extra "capacity" field which says what the true size of the memory allocation is. So yes, you need to implement it this way for full fidelity.

> Second q: this sounds reasonable, but should I be worried that comparing symints every time someone calls .sym_sizes() will cause problems and add a bunch of unnecessary guards? Especifically if the symints are unbacked -we don't actually want to compare the symints, we just want to test that their identity is the same. Is there a way to do that?

You cannot do regular equality. You need to do object identity on the SymNode pointer. BTW, this means we will keep refreshing if someone keeps passing us fresh SymNodes every time we query the inner size, but there isn't really any way around that.

> If someone runs a tensor.t_(1, 0), the key thing here is that we are now lazily updating the buffer, right? So this isn't super safe compared to eager mode, where the updating would happen eagerly during the t_() call?

I guess? But I would argue that it's UB to hold onto the IntArrayRef past a metadata mutating call. Like, maybe you can do it and it works but you shouldn't. Do do it eagerly, you would need to interpose on all metadata updating operations and refresh your buffer.

**Comment by bdhirsh** [comment]
Update: the high-fidelity thing (having the buffer start out with capacity of 5, only reallocating once we hit a size > 5, and realloacating on every future resize) looks like it was needed. I had some test segfaulting due to size calls returning corrupted data, that were fixed when I did the high fidelity thing.

Some other tests are still failing that seem unrelated, I'll do a large push once I finish fixing those

**Comment by bdhirsh** [comment]
Using this PR as a bit of a scratchpad for now to get green CI + debug the XLA CI failures.

**Comment by bdhirsh** [comment]
Ok, I managed to repro the XLA failure in the CI machine (thanks to @JackCaoG - he recommended I try running the repro command with `LD_LIBRARY_PATH="$LD_LIBRARY_PATH:$(python3-config --prefix)/lib"`). After some poking around, I think https://github.com/pytorch/xla/pull/5600 should fix the issue.

I'll kick off another round of CI with the xla commit pinned to https://github.com/pytorch/xla/pull/5600 to confirm.

Assuming that this fixes the problem, then remaining work to make this landable is:

(1) land all the PR's below this in the stack (some have been approved, some are still going through a few rounds of code review with Richard)

(2) There is one final failure (I think? Need to actually re-run CI to confirm that this is the last failure) - which is that some of the AOTExport expecttests have extra `detach()` nodes in them. Need to investigate why.

**Comment by bdhirsh** [comment]
Starting to look at the detach() issue now. Repro:
```
def f(a):
      return a.detach()
```

Tracing this function with C++ vs python functionalization gives:

```
# CPP
def forward(self, arg0_1):
    detach = torch.ops.aten.detach.default(arg0_1);  arg0_1 = None
    return detach

# Python
def forward(self, arg0_1):
    detach = torch.ops.aten.detach.default(arg0_1);  arg0_1 = None
    detach_1 = torch.ops.aten.detach.default(detach);  detach = None
    return detach_1
```

File: .github/ci_commit_pins/xla.txt [hunk 1]
```diff
@@ -1 +1 @@
-fc48cfe24eb99b883585ca16590a3ef32652300e
```
Comment (bdhirsh) [inline .github/ci_commit_pins/xla.txt:1]: I need the pin update because this change requires https://github.com/pytorch/xla/pull/5600

**Comment by bdhirsh** [comment]
@ezyang one of the last hang-ups for python functionalization that I've been looking into is that .detach() calls show up twice in the graph (as seen above).

I know why now, although I'm wondering what you think - I'm currently leaning towards accepting this difference (and leaving it to the downstream users of AOTAutograd to do something reasonable with the double detach, since e.g. inductor will optimize it away pretty painlessly).

Root cause:

(1) ADInplaceOrView ends up calling detach() twice. Once [here](https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/VariableTypeManual.cpp#L445) as a direct dispatcher op, and a second time in the [as_view](https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/VariableTypeManual.cpp#L451) call, which eventually calls [shallow_copy_and_detach()](https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/variable.h#L740)

(2) With python functionalization, we'll eventually trace both detach calls, while with C++ functionalization, we will not add a detach() into the graph on the second call, shallow_copy_and_detach().

Why? FunctionalTensorWrapper overrides shallow_copy_and_detach [here](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/FunctionalTensorWrapper.cpp#L316). It actually does something pretty similar to the vanilla TensorImpl::shallow_copy_and_detach_core, though: It only sends the detach() to a tensor subclass if it key_set_.has(c10::DispatchKey::Python) (and Python is not in the TLS exclude set).

At first I thought this would be the case if we have a FunctionalTensorWrapper holding a FakeTensor, since FunctionalTensorWrapper copies all of the keys from the inner tensor to the wrapper. This is explicitly not true for the Python key though ([here](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/FunctionalTensorWrapper.cpp#L40)), because well, we don't want FunctionalTensorWrapper to ever hit the python key directly.

Based on the above, it seems to me like it would technically be more correct for C++ functionalization to hit the detach() twice, mirroring the behavior of autograd (which technically calls detach twice). If we really wanted to, we could probably arrange for the FunctionalTensor subclass to get an override for shallow_copy_and_detach_core that explicitly says not to redispatch, which would fix this discrepency.

**Comment by ezyang** [comment]
Just accept the difference lol
